src/synapse_mcp/session_storage/redis_backend.py

```diff
@@ -0,0 +1,233 @@
+"""Redis-backed session storage for production deployments."""
+
+import json
+import logging
+from datetime import datetime, timedelta, timezone
+from typing import Optional, Set
+
+from .base import SessionStorage
+
+logger = logging.getLogger("synapse_mcp.session_storage")
+
+try:  # pragma: no cover - import guarded by runtime availability
+    import redis.asyncio as redis
+    REDIS_AVAILABLE = True
+except ImportError:  # pragma: no cover - exercised when redis is absent
+    redis = None  # type: ignore
+    REDIS_AVAILABLE = False
+    logger.warning("Redis not available - falling back to in-memory storage")
+
+
+class RedisSessionStorage(SessionStorage):
+    """Redis-based session storage for production deployments."""
+
+    def __init__(self, redis_url: str, key_prefix: str = "synapse_mcp:session") -> None:
+        if not REDIS_AVAILABLE:
+            raise RuntimeError("Redis dependency not available")
+
+        self.redis_url = redis_url
+        self.key_prefix = key_prefix
+        self._redis: Optional["redis.Redis"] = None
+
+        self.user_subjects_set = f"{key_prefix}:subjects"
+        self.known_tokens_set = f"{key_prefix}:tokens"
+
+    def _subject_token_key(self, user_subject: str) -> str:
+        return f"{self.key_prefix}:user:{user_subject}"
+
+    def _token_subject_key(self, access_token: str) -> str:
+        return f"{self.key_prefix}:token:{access_token}"
+
+    def _token_metadata_key(self, access_token: str) -> str:
+        return f"{self.key_prefix}:metadata:{access_token}"
+
+    async def _get_redis(self) -> "redis.Redis":
+        if self._redis is None:
+            try:
+                self._redis = redis.from_url(
+                    self.redis_url,
+                    decode_responses=True,
+                    socket_connect_timeout=5,
+                    socket_timeout=5,
+                    health_check_interval=30,
+                )
+                await self._redis.ping()
+                logger.info("Redis connection established")
+            except Exception as exc:  # pragma: no cover - connection failure
+                logger.error("Failed to connect to Redis: %s", exc)
```

Comment:
In addition to setting `exc_info=True` in the log message, using `.exception` will do that automatically and emit the log at Error level.

```suggestion
                logger.exception("Failed to connect to Redis")
```